pkg/controller/service/controller_test.go

```diff
@@ -1428,3 +1429,72 @@ func Test_getNodeConditionPredicate(t *testing.T) {
 		})
 	}
 }
+
+func TestListWithPredicate(t *testing.T) {
+	fakeInformerFactory := informers.NewSharedInformerFactory(&fake.Clientset{}, 0*time.Second)
+	var nodes []*v1.Node
+	for i := 0; i < 5; i++ {
+		var phase v1.NodePhase
+		if i%2 == 0 {
+			phase = v1.NodePending
+		} else {
+			phase = v1.NodeRunning
+		}
+		node := &v1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("node-%d", i),
+			},
+			Status: v1.NodeStatus{
+				Phase: phase,
+			},
+		}
+		nodes = append(nodes, node)
+		fakeInformerFactory.Core().V1().Nodes().Informer().GetStore().Add(node)
+	}
+
+	tests := []struct {
+		name      string
+		predicate NodeConditionPredicate
+		expect    []*v1.Node
+	}{
+		{
+			name: "ListWithPredicate filter Running node",
+			predicate: func(node *v1.Node) bool {
+				return node.Status.Phase == v1.NodeRunning
+			},
+			expect: []*v1.Node{nodes[1], nodes[3]},
+		},
+		{
+			name: "ListWithPredicate filter Pending node",
+			predicate: func(node *v1.Node) bool {
+				return node.Status.Phase == v1.NodePending
+			},
+			expect: []*v1.Node{nodes[0], nodes[2], nodes[4]},
+		},
+		{
+			name: "ListWithPredicate filter Terminated node",
+			predicate: func(node *v1.Node) bool {
+				return node.Status.Phase == v1.NodeTerminated
+			},
+			expect: []*v1.Node{},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			get, err := ListWithPredicate(fakeInformerFactory.Core().V1().Nodes().Lister(), test.predicate)
+			sort.Slice(get, func(i, j int) bool {
+				return get[i].Name < get[j].Name
+			})
+			if err != nil {
+				t.Errorf("Error from ListWithPredicate: %v", err)
+			} else if !reflect.DeepEqual(get, test.expect) {
+				// reflect.DeepEqual compares two empty slices will return false.
```

Comment:
> reflect.DeepEqual compares two empty slices will return false

that doesn't seem accurate... empty and nil will return false... the checks done inside this block aren't a very strong comparison, they just make sure the lengths are the same

Reply:
Um...I get it, thank you.

Reply:
addressed